llm/server.go

```diff
@@ -214,15 +214,23 @@ func NewLlamaServer(gpus discover.GpuInfoList, model string, ggml *GGML, adapter
 		params = append(params, "--threads", strconv.Itoa(defaultThreads))
 	}
 
-	flashAttnEnabled := envconfig.FlashAttention()
+	// Check flash attention support and determine KV cache type
```

Comment:
I believe this can be simplified if we keep the flash attention check the same as before. This will also help move flash attention logic outside of `kv_cache.go`

```
flashAttnEnabled := envconfig.FlashAttention()
for _, g := range gpus {
    // only cuda (compute capability 7+) and metal support flash attention
    if g.Library != "metal" && (g.Library != "cuda" || g.DriverMajor < 7) {
	flashAttnEnabled = false
    }

    // mmap has issues with partial offloading on metal
    if g.Library == "metal" &&
        uint64(opts.NumGPU) > 0 &&
	uint64(opts.NumGPU) < ggml.KV().BlockCount()+1 {
	opts.UseMMap = new(bool)
	*opts.UseMMap = false
    }
}

if !modelSupportsFlashAttention(ggml) {
    flashAttnEnabled = false
}

if flashAttnEnabled {
    params = append(params, "--flash-attn")

    // flash attention also supports k/v cache quantization
    kvCacheType := kVCacheQuantization(ggml, gpus)
    if kvCacheType != "" {
        slog.Debug("setting cache type", "type", kvCacheType)
        params = append(params, "--kv-cache-type", kvCacheType)
    }
}

Reply:
I think I've come up with a cleaner / safer solution, have a look at server.go - modelSupportsFlashAttention and checkFlashAttentionSupport

Reply:
I think we can avoid it by adding print statements in the approach above, e.g.

```
if flashAttnEnabled && g.Library != "metal" && (g.Library != "cuda" || g.DriverMajor < 7) {
    slog.Warn("gpu does not support flash attention")
    flashAttnEnabled = false
}

if flashAttnEnabled && !modelSupportsFlashAttention(ggml) {
   slog.Warn("model does not support flash attention")
    flashAttnEnabled = false
}
```

Reply:
OK but this will mean copying all that same logic into memory.go, otherwise the memory estimates will be wrong.

The alternative would be additional parameters to EstimateGPULayers that every caller will have to pass through - which seems worse.

I'll change it all again now.

Reply:
Sorry you're right, a bit of duplication in Go is okay but that would be too much (checking GPU versions etc). Let me review closer